internal/controllers/clusterclass/clusterclass_controller.go

```diff
@@ -133,14 +149,57 @@ func (r *Reconciler) reconcile(ctx context.Context, clusterClass *clusterv1.Clus
 			continue
 		}
 
+		reconciledRefs.Insert(uniqueKey)
+
+		// Add the ClusterClass as owner reference to the templates so clusterctl move
+		// can identify all related objects and Kubernetes garbage collector deletes
+		// all referenced templates on ClusterClass deletion.
 		if err := r.reconcileExternal(ctx, clusterClass, ref); err != nil {
 			errs = append(errs, err)
 			continue
 		}
-		reconciledRefs.Insert(uniqueKey)
+
+		// Check if the template reference is outdated, i.e. it is not using the latest apiVersion
+		// for the current CAPI contract.
+		updatedRef := ref.DeepCopy()
+		if err := conversion.UpdateReferenceAPIContract(ctx, r.Client, r.APIReader, updatedRef); err != nil {
+			errs = append(errs, err)
+		}
+		if ref.GroupVersionKind().Version != updatedRef.GroupVersionKind().Version {
+			outdatedRefs[ref] = updatedRef
+		}
+	}
+	if len(errs) > 0 {
+		return ctrl.Result{}, kerrors.NewAggregate(errs)
+	}
+
+	reconcileConditions(clusterClass, outdatedRefs)
+
+	return ctrl.Result{}, nil
+}
+
+func reconcileConditions(clusterClass *clusterv1.ClusterClass, outdatedRefs map[*corev1.ObjectReference]*corev1.ObjectReference) {
+	if len(outdatedRefs) > 0 {
+		var msg []string
+		for currentRef, updatedRef := range outdatedRefs {
+			msg = append(msg, fmt.Sprintf("%s should be %s", currentRef.GroupVersionKind().String(), updatedRef.GroupVersionKind().String()))
+		}
+		conditions.Set(
+			clusterClass,
+			conditions.FalseCondition(
+				clusterv1.ClusterClassRefVersionsUpToDateCondition,
+				clusterv1.ClusterClassOutdatedRefVersionsReason,
+				clusterv1.ConditionSeverityWarning,
+				strings.Join(msg, ", "),
```

Comment:
How about using a different seperator like `;` (or something else). GVK when converted to string already has `,` in it. The output string could be confusing.


Also, the message only uses the GVK of the ref which does not include the name. If a few DockerMachineTemplate refs are outdated that tt could be confusing if the mesage just repeats that a DockerMachineTemplate ref is outdated without pointing to the correct ref.

Reply:
I changed it to use a quoted string. Keep using `,` as separater seems slight more grammatically correct. The condition now looks like this:
```yaml
  status:
    conditions:
    - lastTransitionTime: "2022-10-07T10:04:22Z"
      message: Ref "infrastructure.cluster.x-k8s.io/v1alpha4, Kind=DockerClusterTemplate
        default/quick-start-my-cluster" should be "infrastructure.cluster.x-k8s.io/v1beta1,
        Kind=DockerClusterTemplate default/quick-start-my-cluster", Ref "infrastructure.cluster.x-k8s.io/v1alpha4,
        Kind=DockerMachineTemplate default/quick-start-control-plane" should be "infrastructure.cluster.x-k8s.io/v1beta1,
        Kind=DockerMachineTemplate default/quick-start-control-plane", Ref "infrastructure.cluster.x-k8s.io/v1alpha4,
        Kind=DockerMachineTemplate default/quick-start-docker-worker-machinetemplate"
        should be "infrastructure.cluster.x-k8s.io/v1beta1, Kind=DockerMachineTemplate
        default/quick-start-docker-worker-machinetemplate"
      reason: OutdatedRefVersions
      severity: Warning
      status: "False"
      type: RefVersionsUpToDate
```
It's a bit verbose, but also very explicit which is probably a good thing